rust_search/src/main.rs

```diff
@@ -0,0 +1,293 @@
+        result.extend([before, "<b>", q, "</b>", after]);
+    } else {
+        result.push_str(text);
+    }
+}
+
+// Ignored as writing to a `String` is always Ok
+#[allow(unused_must_use)]
+fn write_map(result: &mut String, fields: &HashMap<String, Value>) {
+    result.push('{');
+    let mut fields_iter = fields.into_iter();
+    if let Some((key, value)) = fields_iter.next() {
+        write!(result, "\"{}\":", key);
+        write_value(result, value);
+        for (key, value) in fields_iter {
+            write!(result, ",\"{}\":", key);
+            write_value(result, value);
+        }
+    }
+    result.push('}');
+}
+
+// Ignored as writing to a `String` is always Ok
+#[allow(unused_must_use)]
+fn write_value(result: &mut String, value: &Value) {
+    match &value.kind {
+        Some(Kind::DoubleValue(v)) => {
+            write!(result, "{}", v);
+        }
+        Some(Kind::IntegerValue(v)) => {
+            write!(result, "{}", v);
+        }
+        Some(Kind::StringValue(v)) => {
+            write!(result, "\"{}\"", v);
+        }
+        Some(Kind::BoolValue(v)) => {
+            write!(result, "{}", v);
+        }
+        Some(Kind::StructValue(Struct { fields })) => write_map(result, fields),
+        Some(Kind::ListValue(ListValue { values })) => {
+            result.push('[');
+            let mut values_iter = values.into_iter();
+            if let Some(value) = values_iter.next() {
+                write_value(result, value);
+                for value in values_iter {
+                    result.push(',');
+                    write_value(result, value);
+                }
+            }
+            result.push(']');
+        }
+        _ => result.push_str("null"),
+    }
+}
+
+fn add_point(result: &mut String, payload: &HashMap<String, Value>, q: &str) {
+    result.push_str("{\"payload:");
+    write_map(result, &payload);
+    result.push_str(",\"highlight\":\"");
+    if let Some(Kind::StringValue(text)) = &payload.get("text").and_then(|v| v.kind.as_ref()) {
+        highlight(result, &take_n_chars(text, TEXT_LIMIT), q);
+    }
+    result.push_str("\"}");
+}
+
+fn prefix_to_id(prefix: &str) -> PointId {
+    let len = prefix.len();
+    PointId::from(u64::from_le_bytes(if len < 8 {
+        let mut result = [0_u8; 8];
+        result[..len].copy_from_slice(prefix.as_bytes());
+        result
+    } else {
+        prefix.as_bytes()[..8].try_into().unwrap()
+    }))
+}
+
+#[derive(Deserialize)]
+struct Search {
+    q: String,
+    section: String,
+}
+
+#[get("/search")]
+async fn query(
+    context: Data<(BertTokenizer, Session, QdrantClient)>,
+    search: Query<Search>,
+) -> impl Responder {
+    let Search { q, section } = search.into_inner();
+    let (tokenizer, session, qdrant) = context.get_ref();
+    let mut result = String::from("[");
+    let points = if q.len() < 5 {
+        let mut must = vec![Condition::matches("text", q.clone())];
```

Comment:
this condition will exclude everything except for exact text match.

it should be

```suggestion
        let mut must = vec![Condition::matches("text", MatchValue::Text(q.clone()))];
```

Reply:
Actually, the original code is a bit complicated. It falls-back to search without the filter. It can be implemented as batch request for speed

Reply:
This is no longer needed because the text is now stored in the ID to reduce search overhead.